In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('Dataset/Titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.shape

(891, 12)

### Basic cleaning

In [4]:
df.duplicated().any()

False

##### There are no duplicate entries in the data

In [5]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

##### Here the features 'Passerger Id', 'Name' and 'Ticket' have all values unique and are not required for any analysis and thus can be dropped from the data.

In [6]:
df.drop(['PassengerId','Name','Ticket'],axis=1,inplace=True)

In [7]:
df.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64

##### The feature 'Cabin' has most of its entries empty and thus can be dropped.

In [8]:
df.drop('Cabin',axis=1,inplace=True)

##### The features 'Age' and "Embarked' have missing values that can filled with respective mode values.

### Splitting into training and testing data

In [9]:
data=df.values
data

array([[0, 3, 'male', ..., 0, 7.25, 'S'],
       [1, 1, 'female', ..., 0, 71.2833, 'C'],
       [1, 3, 'female', ..., 0, 7.925, 'S'],
       ...,
       [0, 3, 'female', ..., 2, 23.45, 'S'],
       [1, 1, 'male', ..., 0, 30.0, 'C'],
       [0, 3, 'male', ..., 0, 7.75, 'Q']], dtype=object)

In [10]:
X=data[:,1:]
y=data[:,0]

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=213)

In [13]:
X_train.shape

(712, 7)

In [14]:
X_test.shape

(179, 7)

##### There are 712 instances in the training data and 179 instances in test data.

### Missing value handling

In [15]:
from sklearn.impute import SimpleImputer

In [16]:
si=SimpleImputer(missing_values=np.nan,strategy='most_frequent')

In [17]:
si.fit(X_train[:,[2,-1]])

SimpleImputer(strategy='most_frequent')

In [18]:
X_train[:,[2,-1]]=si.transform(X_train[:,[2,-1]])

In [19]:
X_test[:,[2,-1]]=si.transform(X_test[:,[2,-1]])

##### The imputer function was trained using the training data to fill missing values using the mode of respective features and missing values were filled for both train and test data.

### Encoding categorical features

In [20]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

##### Encoding 'Sex' and 'Embarked' columns using one hot encoder

In [21]:
ct = ColumnTransformer(transformers=[('encode', OneHotEncoder(),[1,-1])], remainder='passthrough')

In [22]:
ct.fit(X_train)

ColumnTransformer(remainder='passthrough',
                  transformers=[('encode', OneHotEncoder(), [1, -1])])

In [23]:
X_train=ct.transform(X_train)

In [24]:
X_test=ct.transform(X_test)

### Feature scaling

In [25]:
from sklearn.preprocessing import MinMaxScaler

In [26]:
scaler=MinMaxScaler()

In [27]:
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [28]:
scaler.fit(X_train)

MinMaxScaler()

In [29]:
X_train=scaler.transform(X_train)

In [30]:
X_test=scaler.transform(X_test)

##### Normalized the data using min max scaler